File: src/sagacraft/core/system_registry.py
"""Centralized system configuration and registry - enables easy system management."""

from typing import Dict, List, Optional, Any, Type
from dataclasses import dataclass, field
from enum import Enum

from sagacraft.core.system_base import (
    GameSystem,
    SystemConfig,
    SystemType,
    SystemFactory,
)


@dataclass
class FeatureFlag:
    """Feature flag for enabling/disabling systems."""
    name: str
    enabled: bool = True
    metadata: Dict[str, Any] = field(default_factory=dict)


class SystemRegistry:
    """
    Centralized registry for managing all game systems.
    Single point of truth for system state and configuration.
    """

    def __init__(self):
        self.factory = SystemFactory()
        self.feature_flags: Dict[str, FeatureFlag] = {}
        self.system_configs: Dict[str, SystemConfig] = {}
        self.system_dependencies: Dict[str, List[str]] = {}
        self._initialized = False

    def register_system_class(self, system_id: str, system_class: Type[GameSystem]) -> None:
        """Register a system class for later instantiation."""
        self.factory.register(system_id, system_class)

    def define_system(
        self,
        system_id: str,
        system_type: SystemType,
        enabled: bool = True,
        priority: int = 0,
        dependencies: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Define a system configuration."""
        config = SystemConfig(
            system_id=system_id,
            system_type=system_type,
            enabled=enabled,
            priority=priority,
            metadata=metadata or {},
        )
        self.system_configs[system_id] = config
        self.system_dependencies[system_id] = dependencies or []

        # Create feature flag
        self.feature_flags[system_id] = FeatureFlag(name=system_id, enabled=enabled)

    def enable_system(self, system_id: str) -> None:
        """Enable a system."""
        if system_id in self.feature_flags:
            self.feature_flags[system_id].enabled = True
            if system := self.factory.get(system_id):
                system.enable()

    def disable_system(self, system_id: str) -> None:
        """Disable a system."""
        if system_id in self.feature_flags:
            self.feature_flags[system_id].enabled = False
            if system := self.factory.get(system_id):
                system.disable()

    def is_enabled(self, system_id: str) -> bool:
        """Check if system is enabled."""
        flag = self.feature_flags.get(system_id)
        return flag.enabled if flag else False

    def create_system(self, system_id: str) -> Optional[GameSystem]:
        """Create and register a system instance."""
        if system_id not in self.system_configs:
            raise ValueError(f"System {system_id} not configured")

        config = self.system_configs[system_id]
        return self.factory.create(system_id, config)

    def initialize_all(self, system_ids: Optional[List[str]] = None) -> None:
        """Initialize all or specific systems in dependency order."""
        systems_to_init = system_ids or list(self.system_configs.keys())

        # Sort by priority (higher first)
        systems_to_init.sort(
            key=lambda sid: self.system_configs[sid].priority,
            reverse=True
        )

        for system_id in systems_to_init:
            if not self.is_enabled(system_id):
                continue

            # Check dependencies
            deps = self.system_dependencies.get(system_id, [])
            for dep in deps:
                if dep not in self.factory.systems:
                    raise RuntimeError(f"Dependency {dep} not initialized for {system_id}")

            # Create if not exists
            if self.factory.get(system_id) is None:
                self.create_system(system_id)

        self._initialized = True

    def shutdown_all(self) -> None:
        """Shutdown all systems."""
        self.factory.shutdown_all()
        self._initialized = False

    def get_system(self, system_id: str) -> Optional[GameSystem]:
        """Get a system instance."""
        return self.factory.get(system_id)

    def get_systems_by_type(self, system_type: SystemType) -> List[GameSystem]:
        """Get all systems of a specific type."""
        return self.factory.get_by_type(system_type)

    def get_all_systems(self) -> Dict[str, GameSystem]:
        """Get all systems."""
        return self.factory.get_all()

    def validate_dependencies(self, system_id: str) -> bool:
        """Validate that all dependencies are initialized."""
        deps = self.system_dependencies.get(system_id, [])
        for dep in deps:
            if self.factory.get(dep) is None:
                return False
        return True


# Global registry instance
_global_registry: Optional[SystemRegistry] = None


def get_registry() -> SystemRegistry:
    """Get or create global system registry."""
    global _global_registry
    if _global_registry is None:
        _global_registry = SystemRegistry()
    return _global_registry


def setup_default_systems() -> SystemRegistry:
    """Setup default game systems configuration."""
    registry = get_registry()

    # Core progression systems
    registry.define_system("difficulty", SystemType.GAMEPLAY, priority=100)
    registry.define_system("progression", SystemType.PROGRESSION, priority=90)
    registry.define_system("skill_trees", SystemType.PROGRESSION, priority=85, dependencies=["progression"])
    registry.define_system("battle_pass", SystemType.PROGRESSION, priority=80, dependencies=["progression"])
    registry.define_system("prestige", SystemType.PROGRESSION, priority=75, dependencies=["progression"])

    # Gameplay systems
    registry.define_system("game_modes", SystemType.GAMEPLAY, priority=70)
    registry.define_system("crafting", SystemType.GAMEPLAY, priority=65)
    registry.define_system("environmental", SystemType.GAMEPLAY, priority=60)
    registry.define_system("dungeon_scaling", SystemType.GAMEPLAY, priority=55)

    # Social systems
    registry.define_system("guilds", SystemType.SOCIAL, priority=50)
    registry.define_system("economy", SystemType.SOCIAL, priority=45)
    registry.define_system("mentorship", SystemType.SOCIAL, priority=40)
    registry.define_system("pvp_arenas", SystemType.SOCIAL, priority=35)

    # Content systems
    registry.define_system("daily_challenges", SystemType.CONTENT, priority=30)
    registry.define_system("seasonal", SystemType.CONTENT, priority=25)
    registry.define_system("cosmetics", SystemType.CONTENT, priority=20)
    registry.define_system("achievements", SystemType.CONTENT, priority=15, dependencies=["progression"])

    # Infrastructure systems
    registry.define_system("cloud_saves", SystemType.INFRASTRUCTURE, priority=10)
    registry.define_system("analytics", SystemType.INFRASTRUCTURE, priority=5)

    return registry


class SystemValidator:
    """Validates system configurations and state."""

    @staticmethod
    def validate_system_id(system_id: str) -> bool:
        """Validate system ID format."""
        return bool(system_id) and system_id.replace("_", "").isalnum()

    @staticmethod
    def validate_priority(priority: int) -> bool:
        """Validate priority value."""
        return 0 <= priority <= 1000

    @staticmethod
    def validate_configuration(registry: SystemRegistry, system_id: str) -> tuple[bool, str]:
        """Validate system configuration."""
        if system_id not in registry.system_configs:
            return False, f"System {system_id} not configured"

        config = registry.system_configs[system_id]

        if not SystemValidator.validate_system_id(system_id):
            return False, f"Invalid system ID: {system_id}"

        if not SystemValidator.validate_priority(config.priority):
            return False, f"Invalid priority: {config.priority}"

        # Check dependencies exist
        for dep in registry.system_dependencies.get(system_id, []):
            if dep not in registry.system_configs:
                return False, f"Dependency {dep} not configured"

        return True, "Configuration valid"

    @staticmethod
    def validate_all(registry: SystemRegistry) -> tuple[bool, List[str]]:
        """Validate entire registry configuration."""
        errors = []

        for system_id in registry.system_configs:
            valid, msg = SystemValidator.validate_configuration(registry, system_id)
            if not valid:
                errors.append(msg)

        return len(errors) == 0, errors

File: rust/sagacraft_rs/src/pyport/core/system_registry.rs
use serde_json::Value;
use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SystemType {
    Progression,
    Gameplay,
    Social,
    Content,
    Infrastructure,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SystemConfig {
    pub system_id: String,
    pub system_type: SystemType,
    pub enabled: bool,
    pub priority: i32,
    pub metadata: HashMap<String, Value>,
}

impl SystemConfig {
    pub fn new(system_id: impl Into<String>, system_type: SystemType) -> Self {
        Self {
            system_id: system_id.into(),
            system_type,
            enabled: true,
            priority: 0,
            metadata: HashMap::new(),
        }
    }
}

pub trait GameSystem: 'static {
    fn id(&self) -> &str;
    fn system_type(&self) -> SystemType;
    fn enabled(&self) -> bool;

    fn enable(&mut self);
    fn disable(&mut self);

    fn initialize(&mut self) -> Result<(), String>;
    fn validate(&self) -> bool;
}

type SystemCtor = Box<dyn Fn(SystemConfig) -> Box<dyn GameSystem>>;

pub struct SystemFactory {
    ctors: HashMap<String, SystemCtor>,
    systems: HashMap<String, Box<dyn GameSystem>>,
}

impl SystemFactory {
    pub fn new() -> Self {
        Self {
            ctors: HashMap::new(),
            systems: HashMap::new(),
        }
    }

    pub fn register(&mut self, system_id: impl Into<String>, ctor: SystemCtor) {
        self.ctors.insert(system_id.into(), ctor);
    }

    pub fn create(&mut self, system_id: &str, config: SystemConfig) -> Result<(), String> {
        let ctor = self
            .ctors
            .get(system_id)
            .ok_or_else(|| format!("system not registered: {system_id}"))?;
        let mut sys = ctor(config);
        sys.initialize()?;
        self.systems.insert(system_id.to_string(), sys);
        Ok(())
    }

    pub fn get(&self, system_id: &str) -> Option<&dyn GameSystem> {
        self.systems.get(system_id).map(|s| s.as_ref())
    }

    pub fn get_mut(&mut self, system_id: &str) -> Option<&mut (dyn GameSystem + '_)> {
        self.systems.get_mut(system_id).map(move |s| s.as_mut())
    }

    pub fn shutdown_all(&mut self) {
        self.systems.clear();
    }

    pub fn get_by_type(&self, system_type: SystemType) -> Vec<&dyn GameSystem> {
        self.systems
            .values()
            .filter_map(|s| {
                if s.system_type() == system_type {
                    Some(s.as_ref())
                } else {
                    None
                }
            })
            .collect()
    }

    pub fn get_all(&self) -> HashMap<String, &dyn GameSystem> {
        self.systems
            .iter()
            .map(|(k, v)| (k.clone(), v.as_ref()))
            .collect()
    }

    pub fn has_system(&self, system_id: &str) -> bool {
        self.systems.contains_key(system_id)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct FeatureFlag {
    pub name: String,
    pub enabled: bool,
    pub metadata: HashMap<String, Value>,
}

pub struct SystemRegistry {
    pub factory: SystemFactory,
    pub feature_flags: HashMap<String, FeatureFlag>,
    pub system_configs: HashMap<String, SystemConfig>,
    pub system_dependencies: HashMap<String, Vec<String>>,
    initialized: bool,
}

impl SystemRegistry {
    pub fn new() -> Self {
        Self {
            factory: SystemFactory::new(),
            feature_flags: HashMap::new(),
            system_configs: HashMap::new(),
            system_dependencies: HashMap::new(),
            initialized: false,
        }
    }

    pub fn define_system(
        &mut self,
        system_id: impl Into<String>,
        system_type: SystemType,
        enabled: bool,
        priority: i32,
        dependencies: Vec<String>,
        metadata: HashMap<String, Value>,
    ) {
        let system_id = system_id.into();
        let mut config = SystemConfig::new(system_id.clone(), system_type);
        config.enabled = enabled;
        config.priority = priority;
        config.metadata = metadata;

        self.system_configs.insert(system_id.clone(), config);
        self.system_dependencies
            .insert(system_id.clone(), dependencies);
        self.feature_flags.insert(
            system_id.clone(),
            FeatureFlag {
                name: system_id.clone(),
                enabled,
                metadata: HashMap::new(),
            },
        );
    }

    pub fn enable_system(&mut self, system_id: &str) {
        if let Some(flag) = self.feature_flags.get_mut(system_id) {
            flag.enabled = true;
        }
        if let Some(sys) = self.factory.get_mut(system_id) {
            sys.enable();
        }
    }

    pub fn disable_system(&mut self, system_id: &str) {
        if let Some(flag) = self.feature_flags.get_mut(system_id) {
            flag.enabled = false;
        }
        if let Some(sys) = self.factory.get_mut(system_id) {
            sys.disable();
        }
    }

    pub fn is_enabled(&self, system_id: &str) -> bool {
        self.feature_flags
            .get(system_id)
            .map(|f| f.enabled)
            .unwrap_or(false)
    }

    pub fn create_system(&mut self, system_id: &str) -> Result<(), String> {
        let config = self
            .system_configs
            .get(system_id)
            .cloned()
            .ok_or_else(|| format!("System {system_id} not configured"))?;
        self.factory.create(system_id, config)
    }

    pub fn initialize_all(&mut self, system_ids: Option<Vec<String>>) -> Result<(), String> {
        let mut systems_to_init = system_ids.unwrap_or_else(|| self.system_configs.keys().cloned().collect());

        systems_to_init.sort_by(|a, b| {
            let pa = self.system_configs.get(a).map(|c| c.priority).unwrap_or(0);
            let pb = self.system_configs.get(b).map(|c| c.priority).unwrap_or(0);
            pb.cmp(&pa)
        });

        for system_id in systems_to_init {
            if !self.is_enabled(&system_id) {
                continue;
            }

            let deps = self
                .system_dependencies
                .get(&system_id)
                .cloned()
                .unwrap_or_default();
            for dep in deps {
                if !self.factory.has_system(&dep) {
                    return Err(format!("Dependency {dep} not initialized for {system_id}"));
                }
            }

            if !self.factory.has_system(&system_id) {
                self.create_system(&system_id)?;
            }
        }

        self.initialized = true;
        Ok(())
    }

    pub fn shutdown_all(&mut self) {
        self.factory.shutdown_all();
        self.initialized = false;
    }

    pub fn validate_dependencies(&self, system_id: &str) -> bool {
        let deps = self
            .system_dependencies
            .get(system_id)
            .cloned()
            .unwrap_or_default();
        deps.into_iter().all(|d| self.factory.has_system(&d))
    }

    pub fn is_initialized(&self) -> bool {
        self.initialized
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    struct DummySystem {
        config: SystemConfig,
    }

    impl DummySystem {
        fn new(config: SystemConfig) -> Self {
            Self { config }
        }
    }

    impl GameSystem for DummySystem {
        fn id(&self) -> &str {
            &self.config.system_id
        }

        fn system_type(&self) -> SystemType {
            self.config.system_type.clone()
        }

        fn enabled(&self) -> bool {
            self.config.enabled
        }

        fn enable(&mut self) {
            self.config.enabled = true;
        }

        fn disable(&mut self) {
            self.config.enabled = false;
        }

        fn initialize(&mut self) -> Result<(), String> {
            Ok(())
        }

        fn validate(&self) -> bool {
            true
        }
    }

    #[test]
    fn initialize_all_respects_dependencies() {
        let mut r = SystemRegistry::new();

        r.factory.register(
            "a",
            Box::new(|cfg| Box::new(DummySystem::new(cfg))),
        );
        r.factory.register(
            "b",
            Box::new(|cfg| Box::new(DummySystem::new(cfg))),
        );

        r.define_system(
            "a",
            SystemType::Gameplay,
            true,
            10,
            vec![],
            HashMap::new(),
        );
        r.define_system(
            "b",
            SystemType::Gameplay,
            true,
            9,
            vec!["a".to_string()],
            HashMap::new(),
        );

        // If we request b only, it should error because a not initialized.
        assert!(r.initialize_all(Some(vec!["b".to_string()])).is_err());

        // Initialize a then b works.
        r.shutdown_all();
        r.initialize_all(Some(vec!["a".to_string(), "b".to_string()]))
            .unwrap();
        assert!(r.is_initialized());
        assert!(r.factory.has_system("a"));
        assert!(r.factory.has_system("b"));
    }
}
